### PR TITLE
316 create working active learning example with existing codebase

### PR DESCRIPTION
**Pull request by cisprague**
Added an active learning example --- just stream-based active learning for now.

### CODE REVIEW

**Comment by sgreenbury** [comment]
This is looking great @cisprague - I really like the class structure and experimental set-up exploring the different algorithms! It is also very helpful to see this with the current code base in relation to #322 for the refactor.

A few comments from a first look in case useful:
- This might be very subjective but the term "query" for the method `.query()` returning a `bool` I found slightly unintuitive initially. I think this is since it could be interpreted as a noun i.e. the query to be run (as opposed to the verb whether to run the query on the simulator).
- It might be worth defining query rate precisely (i.e. proportion of queries that that will be run?)
- I like the introduction to the learning algorithms above the subclassing - it might be worth adding a short doc comment to each class describing the algorithm and the parameters to remind the reader as they move through the tutorial of how they relate to each other.
- For the metrics plotted - I wonder if it's worth adding a little more description for what each metric tells us about a learner.  For example: "Uncertainty in the emulator measured with the standard deviation where a greater rate of reduction is indicative of better learner performance since fewer queries are required."
- Related to this - I wonder if the $R^2$ would be worth plotting as metric too?

**Comment by radka-j** [comment]
This looks great! 

I mostly had clarifying questions that @sgreenbury already outlined, it would be useful to just have a bit more detail in the notebook text as well as the docstrings. 

An additional one was that I am unsure about is what membership-based AL strategy is. I know it's not the main focus here but as long as we mention it, is it possible to expand/reword this slightly to make it clearer?

Personally I don't mind the term `query`, especially if it's well established in the active learning domain?

**Comment by radka-j** [comment]
I think we should move all the active learning classes to the codebase and have the notebook import everything from autoemulate and act primarily as a tutorial for active learning, only describing the class structure as required. I guess for now the notebook might need to keep the toy Emulator/Simulator classes. Once we have an Emulator class from the refactor, we can make any small updates to how the active learners use it and the notebook as required.

**Review by radka-j — COMMENTED** [review]
This is really great!!! I learned a lot 😊 

I only had a question about how the rolling window in the r2 computation is implemented. Also, as mentioned above, I'd be in favour of moving everything into the codebase in this PR but am happy if others think that should be handled separately. 

The only last thing is that there are now 2 notebooks here, we probably want to delete the old one?

**Comment by sgreenbury** [comment]
These updates look really great @cisprague! I found the additional technical detail/diagrams all very helpful. I think moving the classes into the library as @radka-j suggests sounds great and will be helpful for revising in the refactor (related to #322, #341, #327). I like the idea for the automatic registry too!

For the $R^2$ plot, the window as a parameter discussed above sounds like it will help so that each learner is averaged on the same number of values making their series more comparable. For the difference in length of the series, perhaps adding another field to the `Stream` to capture the iteration number that each query occurred could be useful (i.e. `self.queried_iteration.append(len(self.queries))`)? This could then be used to place the $R^2$ values at the iteration associated with e.g. the right hand side of the window and the series could extend across all iterations.

Another option might be to optionally add validation/test data into the `Learner` or `Stream` API. Like validation data in [Keras](https://keras.io/api/models/model_training_apis/#fit-method)/[Lightning](https://lightning.ai/docs/pytorch/stable/common/trainer.html#fit). However, this seems beyond the scope of the `Learner` (since it is monitoring performance on additional held-out data) so seems like a less good option here.

Two other small comments:
- I think the following deps (`seaborn` and `anytree`) need to be added to the dev deps to run the notebook?
- We might consider replacing the `List` and `Tuple` type hints with `list` and `tuple` (see [ruff pep585](https://docs.astral.sh/ruff/rules/non-pep585-annotation/)) since we're targeting Python >= 3.10? I can open an issue to consider if additional lints would be helpful to have in the pre-commit / CI.

**Comment by cisprague** [comment]
@radka-j and @sgreenbury, thank you for the comments. The updates in this PR were written against the previous version of the code. The updated tutorial notebook should be put there since the parent PR (#301) refactors the API. Closing the PR.
